tempesta_fw/classifier/req_conn_limit.c

```diff
@@ -364,45 +608,80 @@ rcl_init(void)
 		return -EINVAL;
 	}
 
+	rcl_ctl = register_net_sysctl(&init_net, "net/tempesta/req_conn_limit",
+				      rcl_ctl_table);
+	if (!rcl_ctl) {
+		TFW_ERR("rcl: can't register sysctl table\n");
+		r = -1;
+		goto err_sysctl;
+	}
+
 	r = tfw_classifier_register(&rcl_class_ops);
 	if (r) {
 		TFW_ERR("rcl: can't register classifier\n");
 		goto err_class;
 	}
 
-	r = tfw_gfsm_register_fsm(TFW_FSM_RCL, rcl_http_req_handler);
+	/**
+	 * FIXME:
+	 *  Here we add two primitive hooks by registering two FSMs.
+	 *  There is a bunch of problems here:
+	 *  - We can't unregister hooks. Therefore, we can't unload this module
+	 *    and can't recover if the second tfw_gfsm_register_hook() fails.
+	 *  - We have to add every hook to the global enum of FSMs.
+	 *  - We register two FSMs, but actually don't implement anything close
+	 *    to FSM and don't need the FSM switching logic.
+	 *
+	 * The suggested solution is to extend the GFSM with the support of
+	 * "lightweight hooks" that behave like plain functions rather than FSM.
+	 * That should solve all these problems listed above:
+	 *  - Such hook may be unregistered any time it is not executed.
+	 *  - The hook doesn't need an ID, so no need to maintain a global list
+	 *    of all known hooks in the GFSM code.
+	 *  - No need to create a dummy FSM just to call the hook.
+	 *    This is easy to comprehend, and perhaps faster since the GFSM
+	 *    doesn't need to switch FSMs.
+	 */
```

Comment:
The comment isn't relevant, this is just an improper GFSM usage. GFSM allows you to register a FSM, not just a hook. When you call tfw_gfsm_register_fsm() you register FSM handler which typicaly should run switch for current state of the tracked object. So you shouldn't have been registering the second FSM, but you should register second hook as you do.